### PR TITLE
yaml: phase 2 DBConfigs

### PR DESCRIPTION
**Pull request by sougou**
We continue to support backward compatibility.

### CODE REVIEW

File: go/vt/dbconfigs/dbconfigs.go [hunk 1]
```diff
@@ -255,123 +285,139 @@ func (dbcfgs *DBConfigs) makeParams(userKey string, withDB bool) Connector {
 
 // IsZero returns true if DBConfigs was uninitialized.
 func (dbcfgs *DBConfigs) IsZero() bool {
-	return len(dbcfgs.userConfigs) == 0
+	return *dbcfgs == DBConfigs{}
+}
+
+// HasGlobalSettings returns true if DBConfigs was uninitialized.
+func (dbcfgs *DBConfigs) HasGlobalSettings() bool {
+	return dbcfgs.Host != "" || dbcfgs.Socket != ""
```
Comment (harshit-gangal) [inline go/vt/dbconfigs/dbconfigs.go:293]: did not follow the comment with the check below.

File: go/vt/vtcombo/tablet_map.go [hunk 2]
```diff
@@ -179,9 +179,8 @@ func InitTabletMap(ts *topo.Server, tpb *vttestpb.VTTestTopology, mysqld mysqlct
 					if dbname == "" {
 						dbname = fmt.Sprintf("vt_%v_%v", keyspace, shard)
 					}
-					// Copy dbcfgs and override SidecarDBName because there will be one for each db.
-					copydbcfgs := dbcfgs.Copy()
-					copydbcfgs.SidecarDBName.Set("_" + dbname)
+					// Clone dbcfgs and override SidecarDBName because there will be one for each db.
+					copydbcfgs := dbcfgs.Clone()
```
Comment (harshit-gangal) [inline go/vt/vtcombo/tablet_map.go:183]: Is the cloning required? SidecarDBName setting is removed in the code.

File: go/yaml2/yaml.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package yaml2 ensures that the right yaml package gets imported.
+// The default package that goimports adds is not the one we want to use.
+package yaml2
```
Comment (harshit-gangal) [inline go/yaml2/yaml.go:19]: nice hack.

**Comment by harshit-gangal** [comment]
> We continue to support backward compatibility.

should be we continue to support backward compatibility on deprecated db config flags as well.

File: go/vt/dbconfigs/dbconfigs.go [hunk 4]
```diff
@@ -255,123 +285,139 @@ func (dbcfgs *DBConfigs) makeParams(userKey string, withDB bool) Connector {
...
+
+// Redacted redacts passwords from DBConfigs.
+func (dbcfgs *DBConfigs) Redacted() *DBConfigs {
+	dbcfgs = dbcfgs.Clone()
+	dbcfgs.App.Password = "****"
+	dbcfgs.Dba.Password = "****"
+	dbcfgs.Filtered.Password = "****"
+	dbcfgs.Repl.Password = "****"
+	dbcfgs.Appdebug.Password = "****"
+	dbcfgs.Allprivs.Password = "****"
```
[the remark this answers is not in the document]
Comment (sougou) [reply go/vt/dbconfigs/dbconfigs.go:318]: It's just general safety whether it's SSL or not.

File: go/vt/vtcombo/tablet_map.go (hunk 2 above)
Comment (sougou) [reply go/vt/vtcombo/tablet_map.go:183]: It's safer. I also think the DBName will get set differently for each tablet.

**Pull request merged**